File: ldk-common/src/main/java/com/lgou2w/ldk/common/IteratorChain.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * Copyright (C) 2018-2020 The lgou2w <lgou2w@hotmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lgou2w.ldk.common;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;

/*
 * Modify: Apache common-collections IteratorChain
 * by lgou2w on 09/19/2018
 */

public final class IteratorChain<T> implements Iterator<T> {
  private final List<Iterator<T>> iterators;
  private @Nullable Iterator<T> current;
  private int index = 0;

  @SafeVarargs
  private IteratorChain(Iterator<T>... iterators) {
    this.iterators = Arrays.asList(iterators.clone());
  }

  private IteratorChain(Collection<Iterator<T>> iterators) {
    this.iterators = new ArrayList<>(iterators);
  }

  @NotNull
  private Iterator<T> update() {
    if (current == null) {
      current = iterators.isEmpty() ? (Iterator<T>) EMPTY : iterators.get(0);
    } else {
      while (!current.hasNext() && index < iterators.size() - 1) {
        index += 1;
        current = iterators.get(index);
      }
    }
    return current;
  }

  @Override
  public boolean hasNext() {
    return update().hasNext();
  }

  @Override
  public T next() {
    return update().next();
  }

  private final static Iterator<Object> EMPTY = new Iterator<Object>() {
    @Override
    public boolean hasNext() {
      return false;
    }

    @Override
    public Object next() {
      throw new NoSuchElementException("Empty iterator");
    }

    @Override
    public void remove() {
      throw new UnsupportedOperationException("Empty iterator");
    }
  };

  @NotNull
  @Contract("null -> fail; !null -> new")
  public static <T> IteratorChain<T> concat(Iterator<T>... iterators) {
    if (iterators == null) throw new NullPointerException("iterators");
    return new IteratorChain<>(iterators);
  }

  @NotNull
  @Contract("null -> fail; !null -> new")
  public static <T> IteratorChain<T> concat(Collection<Iterator<T>> iterators) {
    if (iterators == null) throw new NullPointerException("iterators");
    return new IteratorChain<>(iterators);
  }
}

File: ldk-common/src/main/kotlin/com/lgou2w/ldk/common/IteratorChain.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * Copyright (C) 2018 The lgou2w (lgou2w@hotmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lgou2w.ldk.common

/*
 * Modify: Apache common-collections IteratorChain Java -> Kotlin
 * by lgou2w on 19/09/2018
 */

/**
 * ## IteratorChain (迭代链)
 *
 * * An iterator chain separated by Apache's `common-collections` library.
 *      A tool class used to connect multiple iterators.
 * * 由 Apache 的 `common-collections` 库分离出来的迭代链. 用来将多个迭代器连接起来的工具类.
 *
 * ### Sample:
 * ```kotlin
 * val list1 = listOf(0, 1, 2)
 * val list2 = listOf(3, 4, 5)
 * val iterator = IteratorChain.concat(list1.iterator(), list2.iterator(), ...)
 * while (iterator.hasNext())
 *      println(iterator.next()) // 0, 1, 2, 3, 4, 5
 * ```
 *
 * @see [concat]
 * @see [Iterator]
 * @author Apache, lgou2w
 */
class IteratorChain<T> private constructor(
        iterators: Array<out Iterator<T>>
) : Iterator<T> {

    private val iterators : List<Iterator<T>> = iterators.toList()
    private var lastUsedIterator : Iterator<T>? = null
    private var currentIterator : Iterator<T>? = null
    private var currentIndex = 0

    private fun updateCurrentIterator() {
        if (currentIterator == null) {
            @Suppress("UNCHECKED_CAST")
            currentIterator = if (iterators.isEmpty()) EMPTY as Iterator<T> else iterators[0]
            lastUsedIterator = currentIterator
        } else {
            while (!currentIterator.notNull().hasNext() && currentIndex < iterators.size - 1) {
                currentIndex += 1
                currentIterator = iterators[currentIndex]
            }
        }
    }

    override fun hasNext(): Boolean {
        updateCurrentIterator()
        lastUsedIterator = currentIterator
        return currentIterator.notNull().hasNext()
    }

    override fun next(): T {
        updateCurrentIterator()
        lastUsedIterator = currentIterator
        return currentIterator.notNull().next()
    }

    companion object {

        @JvmStatic
        private val EMPTY = object : MutableIterator<Any> {
            override fun hasNext(): Boolean = false
            override fun next(): Any = throw NoSuchElementException("Empty Iterator")
            override fun remove() = throw UnsupportedOperationException("Empty Iterator")
        }

        /**
         * * Joins a given iterator array into an iterator chain object.
         * * 将给定的迭代器数组连接成一个迭代链对象.
         *
         * @param [iterators] Iterator array
         * @param [iterators] 迭代器数组
         */
        @JvmStatic
        fun <T> concat(vararg iterators: Iterator<T>) : IteratorChain<T>
                = IteratorChain(iterators)

        /**
         * * Joins a given iterator collection into an iterator chain object.
         * * 将给定的迭代器集合连接成一个迭代链对象.
         *
         * @param [iterators] Iterator collection
         * @param [iterators] 迭代器集合
         */
        @JvmStatic
        fun <T> concat(iterators: Collection<Iterator<T>>) : IteratorChain<T>
                = IteratorChain(iterators.toTypedArray())
    }
}
